components/kyma-environment-broker/cmd/cli/credential/manager.go

```diff
@@ -0,0 +1,147 @@
+	GetTokenByROPC(ctx context.Context, username, password string) (string, error)
+	TokenExpiry() time.Time
+	Token() (*oauth2.Token, error)
+}
+
+type manager struct {
+	getter *credentialplugin.GetToken
+	input  credentialplugin.Input
+	token  string
+	expiry time.Time
+}
+
+type tokenWriter struct {
+	mgr *manager
+}
+
+func (w *tokenWriter) Write(out credentialpluginwriter.Output) error {
+	w.mgr.cacheToken(out.Token, out.Expiry)
+	return nil
+}
+
+// NewManager Constructs a new credential.Manager using the given OIDC provider and client credentials
+func NewManager(oidcIssuerURL, oidcClientID, oidcClientSecret string, logger logger.Logger) Manager {
+	clock := &clock.Real{}
+	reader := &reader.Reader{}
+	auth := &authentication.Authentication{
+		Clock:  clock,
+		Logger: logger,
+		OIDCClient: &oidcclient.Factory{
+			Clock:  clock,
+			Logger: logger,
+		},
+		AuthCodeBrowser: &authcode.Browser{
+			Logger:  logger,
+			Browser: &browser.Browser{},
+		},
+		AuthCodeKeyboard: &authcode.Keyboard{
+			Logger: logger,
+			Reader: reader,
+		},
+		ROPC: &ropc.ROPC{
+			Logger: logger,
+			Reader: reader,
+		},
+	}
+
+	mgr := &manager{
+		input: credentialplugin.Input{
+			IssuerURL:     oidcIssuerURL,
+			ClientID:      oidcClientID,
+			ClientSecret:  oidcClientSecret,
+			TokenCacheDir: defaultTokenCacheDir,
+		},
+	}
+	writer := &tokenWriter{mgr: mgr}
+	getToken := &credentialplugin.GetToken{
+		Logger:               logger,
+		Authentication:       auth,
+		TokenCacheRepository: &tokencache.Repository{},
+		NewCertPool:          certpool.New,
+		Writer:               writer,
+	}
+	mgr.getter = getToken
+
+	return mgr
+}
+
+// GetToken fetches an ID token from local cache if a valid token is found, or else initiates interactive authorization code grant flow with browser to request a new ID token
+func (mgr *manager) GetTokenByAuthCode(ctx context.Context) (string, error) {
+	in := mgr.input
+	in.GrantOptionSet.AuthCodeBrowserOption = &authcode.BrowserOption{
+		BindAddress:           defaultListenAddress,
+		SkipOpenBrowser:       false,
+		AuthenticationTimeout: time.Duration(defaultAuthenticationTimeoutSec) * time.Second,
+		RedirectURLHostname:   "localhost",
+	}
+	err := mgr.getter.Do(ctx, in)
+	if err != nil {
+		return "", err
+	}
+	return mgr.token, nil
+}
+
+// GetTokenByROPC fetches an ID token from local cache if a valid token is found, or else initiates resource owner password credentials flow to request a new ID token
+func (mgr *manager) GetTokenByROPC(ctx context.Context, username, password string) (string, error) {
+	in := mgr.input
+	in.GrantOptionSet.ROPCOption = &ropc.Option{
+		Username: username,
+		Password: password,
+	}
+	err := mgr.getter.Do(ctx, in)
+	if err != nil {
+		return "", err
+	}
+	return mgr.token, nil
+}
+
+// Token uses GetTokenByAuthCode() to obtain an ID token in oauth2.Token format. This method implements the oauth2.TokenSource interface
+func (mgr *manager) Token() (*oauth2.Token, error) {
```

Comment:
can we use cached token instead of requesting it every time?

Reply:
The token is actually cached by  `credentialplugin.GetToken` and it also ensures refreshing the cached token when it expires. We shouldn't add duplicated logic in our manager related to caching and refresh.